instance/models/utils.py

```diff
@@ -44,3 +55,263 @@ def save(self, *args, **kwargs):
+        feature is not present.
+        """
+        for state in accepted_states:
+            assert state in self.state_classes
+
+        def wrap(method):
+            """
+            Create a MagicWrapper descriptor that will implement the only_for() behavior.
+            """
+            def require_valid_state(resource):
+                """
+                Raise a WrongStateException if resource is not in one of the required states
+                """
+                state = self.__get__(resource)
+                if not isinstance(state, accepted_states):
+                    raise WrongStateException("The method '{}' cannot be called in this state ({} / {}).".format(
+                        method.__name__, state.name, state.__class__.__name__  # pylint: disable=no-member
+                    ))
+
+            descriptor = self
+
+            class MagicWrapper:
+                """ Class which can wrap a method; the result can be used as a property or as a method. """
+                def __call__(self, resource):
+                    """ We are wrapping a property, not a method. No fancy stuff needed. """
+                    require_valid_state(resource)
+                    return method(resource)
+
+                def __get__(self, resource, _type):
+                    """ Get the (wrapped) method, and add a .is_available method to it """
+                    def wrapped_method(*args, **kwargs):
+                        """ Wrapper around the method which checks the state requirements first """
+                        require_valid_state(resource)
+                        return method(resource, *args, **kwargs)
+                    wrapped_method.is_available = lambda: isinstance(descriptor.__get__(resource), accepted_states)
+                    return wrapped_method
+            return MagicWrapper()
+        return wrap
+
+    def transition(self, to_state, from_states=None):
+        """
+        Returns a "transition" method, which can be called to change the current state from
+        state(s) 'from' to 'to'.
+
+        from_states can be a state class, a tuple of classes, a parent class or mixin of various
+        state classes, a tuple of mixins, etc.. if None, any valid state will be accepted.
+        """
+        assert to_state in self.state_classes
+
+        def do_transition(resource):
+            """ The method that performs a specific transition """
+            current_state = self.__get__(resource)
+            if from_states and not isinstance(current_state, from_states):
+                raise WrongStateException("This transition cannot be used to move from {} to {}".format(
+                    current_state.name, to_state.name  # pylint: disable=no-member
+                ))
+            self._set_state(resource, to_state)
+        do_transition.from_states = from_states  # Convenient way for other code to inspect this transition
```

Comment:
@bradenmacdonald nit: if `from_states` is None, this will also be `None`; I believe in such case `self.state_classes` would make a bit more sense.